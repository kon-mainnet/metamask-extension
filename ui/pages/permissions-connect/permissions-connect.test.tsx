import React from 'react';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import { ApprovalType } from '@metamask/controller-utils';
import { BtcAccountType } from '@metamask/keyring-api';
import messages from '../../../app/_locales/en/messages.json';
import { renderWithProvider } from '../../../test/lib/render-helpers';
import mockState from '../../../test/data/mock-state.json';
import { CONNECT_ROUTE } from '../../helpers/constants/routes';
import { createMockInternalAccount } from '../../../test/jest/mocks';
import { shortenAddress } from '../../helpers/utils/util';
import PermissionApprovalContainer from './permissions-connect.container';

const mockPermissionRequestId = '0cbc1f26-8772-4512-8ad7-f547d6e8b72c';

jest.mock('../../store/actions', () => {
  return {
    getRequestAccountTabIds: jest.fn().mockReturnValue({
      type: 'SET_REQUEST_ACCOUNT_TABS',
      payload: {},
    }),
  };
});

const mockAccount = createMockInternalAccount({ name: 'Account 1' });
const mockBtcAccount = createMockInternalAccount({
  name: 'BTC Account',
  address: 'bc1qar0srrr7xfkvy5l643lydnw9re59gtzzwf5mdq',
  type: BtcAccountType.P2wpkh,
});

const defaultProps = {
  history: {
    location: {
      pathname: `${CONNECT_ROUTE}/${mockPermissionRequestId}`,
    },
  },
  location: {
    pathname: `${CONNECT_ROUTE}/${mockPermissionRequestId}`,
  },
  match: {
    params: {
      id: mockPermissionRequestId,
    },
  },
};

const render = (
  props = defaultProps,
  type: ApprovalType = ApprovalType.WalletRequestPermissions,
) => {
  let pendingPermission;
  if (type === ApprovalType.WalletRequestPermissions) {
    pendingPermission = {
      id: mockPermissionRequestId,
      origin: 'https://metamask.io',
      type: ApprovalType.WalletRequestPermissions,
      time: 1721376328642,
      requestData: {
        metadata: {
          id: mockPermissionRequestId,
          origin: 'https://metamask.io',
        },
        permissions: {
          eth_accounts: {},
        },
      },
      requestState: null,
      expectsResult: false,
    };
  }

  const state = {
    ...mockState,
    metamask: {
      ...mockState.metamask,
      internalAccounts: {
        accounts: {
          [mockAccount.id]: mockAccount,
          [mockBtcAccount.id]: mockBtcAccount,
        },
        selectedAccount: mockAccount.id,
      },
      keyrings: [
        {
          type: 'HD Key Tree',
          accounts: [mockAccount.address],
        },
        {
          type: 'Snap Keyring',
          accounts: [mockBtcAccount.address],
        },
      ],
      accounts: {
        [mockAccount.address]: {
          address: mockAccount.address,
          balance: '0x0',
        },
      },
      balances: {
        [mockBtcAccount.id]: {},
      },
      pendingApprovals: {
        [mockPermissionRequestId]: pendingPermission,
      },
    },
  };
  const middlewares = [thunk];
  const mockStore = configureStore(middlewares);
  const store = mockStore(state);

  return renderWithProvider(
    <PermissionApprovalContainer {...props} />,
    store,
    `${CONNECT_ROUTE}/${mockPermissionRequestId}`,
  );
};

describe('PermissionApprovalContainer', () => {
  describe('ConnectPath', () => {
    it('renders correctly', () => {
      const { container, getByText } = render();
      expect(getByText(messages.next.message)).toBeInTheDocument();
      expect(getByText(messages.cancel.message)).toBeInTheDocument();
      expect(container).toMatchSnapshot();
    });

    it('renders the list without BTC accounts', async () => {
      const { getByText, queryByText } = render();
      expect(
        getByText(
          `${mockAccount.metadata.name} (${shortenAddress(
            mockAccount.address,
          )})`,
        ),
      ).toBeInTheDocument();
      expect(
        queryByText(
          `${mockBtcAccount.metadata.name} (${shortenAddress(
            mockBtcAccount.address,
          )})`,
        ),
      ).not.toBeInTheDocument();
    });
  });
});
