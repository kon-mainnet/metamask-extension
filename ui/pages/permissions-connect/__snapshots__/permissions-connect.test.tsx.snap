// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PermissionApprovalContainer ConnectPath renders correctly 1`] = `
<div>
  <div
    class="permissions-connect"
  >
    <div
      class="mm-box"
    >
      <div
        class="mm-box mm-box--padding-4 mm-box--display-flex mm-box--align-items-center mm-box--width-full mm-box--background-color-background-default"
        style="box-shadow: var(--shadow-size-lg) var(--color-shadow-default);"
      >
        <div
          class="mm-box"
        >
          <div
            class="mm-box mm-text mm-avatar-base mm-avatar-base--size-lg mm-text--body-lg-medium mm-text--text-transform-uppercase mm-box--display-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-text-alternative mm-box--background-color-background-alternative mm-box--rounded-full"
            style="border-width: 0px;"
          >
            m
          </div>
        </div>
        <div
          class="mm-box mm-box--margin-right-4 mm-box--margin-left-4 mm-box--display-flex mm-box--flex-direction-column"
          style="overflow: hidden;"
        >
          <p
            class="mm-box mm-text mm-text--body-md mm-text--font-weight-medium mm-text--ellipsis mm-box--color-text-default"
          >
            metamask.io
          </p>
          <p
            class="mm-box mm-text mm-text--body-sm mm-text--ellipsis mm-box--color-text-alternative"
          >
            https://metamask.io
          </p>
        </div>
      </div>
    </div>
    <div
      class="mm-box permissions-connect-choose-account__content mm-box--padding-right-6 mm-box--padding-left-6 mm-box--display-flex mm-box--flex-direction-column mm-box--width-full mm-box--height-full mm-box--background-color-background-alternative"
    >
      <div
        class="mm-box mm-box--padding-top-4 mm-box--padding-bottom-4 mm-box--display-flex mm-box--flex-direction-column mm-box--justify-content-center mm-box--align-items-center"
      >
        <h3
          class="mm-box mm-text mm-text--heading-md mm-box--color-text-default"
        >
          Connect with MetaMask
        </h3>
        <p
          class="mm-box mm-text mm-text--body-md mm-box--color-text-default"
        >
          Select the account(s) to use on this site
        </p>
      </div>
      <div
        class="choose-account-list"
      >
        <div
          class="choose-account-list__header--one-item"
        >
          <button
            class="mm-box mm-text mm-button-base mm-button-link mm-button-link--size-auto mm-text--body-md-medium mm-box--padding-0 mm-box--padding-right-0 mm-box--padding-left-0 mm-box--display-inline-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-info-default mm-box--background-color-transparent"
            style="cursor: pointer;"
          >
            New account
          </button>
        </div>
        <div
          class="choose-account-list__wrapper"
        >
          <div
            class="mm-box choose-account-list__list"
            style="overflow-x: hidden;"
          >
            <div
              class="mm-box choose-account-list__account mm-box--display-flex mm-box--width-full mm-box--background-color-primary-muted"
              data-testid="choose-account-list-0"
            >
              <div
                class="mm-box mm-box--display-flex mm-box--align-items-center mm-box--width-full"
              >
                <label
                  class="mm-box mm-text mm-checkbox mm-text--body-md mm-box--display-inline-flex mm-box--align-items-center mm-box--color-text-default"
                >
                  <span
                    class="mm-checkbox__input-wrapper"
                  >
                    <input
                      checked=""
                      class="mm-box mm-checkbox__input mm-checkbox__input--checked mm-box--margin-0 mm-box--margin-right-0 mm-box--display-flex mm-box--background-color-primary-default mm-box--rounded-sm mm-box--border-color-primary-default mm-box--border-width-2 box--border-style-solid"
                      type="checkbox"
                    />
                    <span
                      class="mm-box mm-checkbox__icon mm-icon mm-icon--size-md mm-box--display-inline-block mm-box--color-primary-inverse"
                      style="mask-image: url('./images/icons/check-bold.svg');"
                    />
                  </span>
                </label>
                <div
                  class="mm-box mm-box--margin-left-2"
                >
                  <div
                    class=""
                  >
                    <div
                      class="identicon"
                      style="height: 34px; width: 34px; border-radius: 17px;"
                    >
                      <div
                        style="border-radius: 50px; overflow: hidden; padding: 0px; margin: 0px; width: 34px; height: 34px; display: inline-block; background: rgb(35, 140, 225);"
                      >
                        <svg
                          height="34"
                          width="34"
                          x="0"
                          y="0"
                        >
                          <rect
                            fill="#FA4300"
                            height="34"
                            transform="translate(-1.8190711089650118 -0.7352934700785319) rotate(264.3 17 17)"
                            width="34"
                            x="0"
                            y="0"
                          />
                          <rect
                            fill="#018E77"
                            height="34"
                            transform="translate(-12.049166321096887 -15.929310915006274) rotate(296.3 17 17)"
                            width="34"
                            x="0"
                            y="0"
                          />
                          <rect
                            fill="#F26E02"
                            height="34"
                            transform="translate(19.34570196808791 -19.44479167700129) rotate(444.6 17 17)"
                            width="34"
                            x="0"
                            y="0"
                          />
                        </svg>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="mm-box mm-box--padding-left-3 mm-box--display-flex mm-box--justify-content-space-between mm-box--width-full"
                  style="min-width: 0;"
                >
                  <div
                    class="mm-box mm-box--display-flex mm-box--flex-direction-column mm-box--width-full"
                  >
                    <p
                      class="mm-box mm-text mm-text--body-md-medium mm-text--ellipsis mm-box--color-text-default"
                      style="text-wrap: nowrap;"
                    >
                      Account 1 (0xd5e09...81111)
                    </p>
                    <div
                      class="mm-box mm-box--display-flex"
                    >
                      <div
                        class="mm-box currency-display-component mm-box--display-flex mm-box--flex-wrap-wrap mm-box--align-items-center"
                        style="flex-wrap: nowrap;"
                        title="0 ETH"
                      >
                        <span
                          class="mm-box mm-text currency-display-component__text mm-text--body-sm mm-text--ellipsis mm-box--color-text-alternative"
                        >
                          0
                        </span>
                        <span
                          class="mm-box mm-text currency-display-component__suffix mm-text--body-sm mm-box--margin-inline-start-1 mm-box--color-text-alternative"
                        >
                          ETH
                        </span>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="mm-box permissions-connect-choose-account__footer mm-box--padding-top-4 mm-box--background-color-background-alternative"
    >
      <div
        class="mm-box mm-box--display-flex mm-box--justify-content-center mm-box--align-items-center"
      >
        <p
          class="mm-box mm-text mm-text--body-md mm-box--color-text-default"
        >
          <span>
             
            Only connect with sites you trust. 
            <button
              class="mm-box mm-text mm-button-base mm-button-link mm-button-link--size-inherit mm-text--body-md-medium mm-box--padding-0 mm-box--padding-right-0 mm-box--padding-left-0 mm-box--display-inline-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-primary-default mm-box--background-color-transparent"
              target="_blank"
            >
              Learn more
            </button>
            
             
          </span>
        </p>
      </div>
      <div
        class="page-container__footer"
      >
        <footer>
          <button
            class="button btn--rounded btn-default page-container__footer-button page-container__footer-button__cancel"
            data-testid="page-container-footer-cancel"
          >
            Cancel
          </button>
          <button
            class="button btn--rounded btn-primary page-container__footer-button"
            data-testid="page-container-footer-next"
          >
            Next
          </button>
        </footer>
      </div>
    </div>
  </div>
</div>
`;
